### DISCUSSION TITLE
`brew doctor` outputs to stderr

### DISCUSSION THREAD

**Discussion by injust** [Homebrew/discussions#5503] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: 1c98a33a376e5b35ebc801d0f7d6ba7b9eb118ac
Last commit: 11 days ago
Core tap JSON: 18 Jul 05:43 UTC
Core cask tap JSON: 18 Jul 05:43 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: micro
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.3.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.3/bin/ruby
CPU: octa-core 64-bit kabylake
Clang: 15.0.0 build 1500
Git: 2.45.2 => /usr/local/bin/git
Curl: 8.6.0 => /usr/bin/curl
macOS: 14.5-x86_64
CLT: 15.3.0.0.1.1708646388
Xcode: N/A
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Putting non-prefixed findutils in your path can cause python builds to fail.

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  fish_add_path /usr/local/sbin
```

### Description of issue

When I was trying to get the output of `brew doctor`, I ran `brew doctor | pbcopy`. This didn't actually copy the output as `brew doctor` outputs to `stderr`.

Is there a specific reason for this? If this command is mostly used for debugging when filing an issue, would it be easier to use if it outputs to `stdout` instead?

**Comment by p-linnane** [comment]
`brew doctor` prints to `stdout` when there are no issues, and `stderr` when there are. This is pretty standard.

```
╰─ brew doctor
Your system is ready to brew.

╰─ echo $?
0

╰─ brew unlink rust
Unlinking /opt/homebrew/Cellar/rust/1.79.0_1... 55 symlinks removed.

╰─ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  rust

╰─ echo $?
1
```

**Reply by injust** [reply]
I can understand why this might seem standard, but it's a bit unintuitive -- I _asked_ for `brew` to print these warnings, so this is expected output.

So I argue that _standard_ is not "errors go to stderr, other output goes to stdout", but instead "expected output goes to stdout, unexpected output goes to stderr".

In that sense, if `brew doctor` is broken and actually outputs an error, then _that_ should go to stderr.

A Python example when type checking using `mypy`: If there are no issues, it exits 0 and prints a success message to stdout. If it finds a syntax error, it exits 2 and prints the errors to stdout. Yes, they are errors, but I _asked_ `mypy` to check for them.

**Reply by injust** [reply]
Also from a usability standpoint, it'd be nice to be able to pipe these to `pbcopy`, a pager, etc.

**Reply by p-linnane** [reply]
You can easily do what you want by running `brew doctor 2>&1 | pbcopy`.

**Reply by injust** [reply]
> You can easily do what you want by running `brew doctor 2>&1 | pbcopy`.

That's what I ended up doing, but IMO it's a workaround for bad default behaviour.

**Reply by injust** [reply]
I'm not sure we really agree on what is correct here, but would you accept a PR to change the output to stdout?

I just can't come up with a scenario where outputting to stderr here would be more useful.

**Reply by p-linnane** [reply]
We're always happy to review PR's and discuss.

**Reply by MikeMcQuaid** [reply, marked as answer]
> but would you accept a PR to change the output to stdout?

No, sorry.

> I just can't come up with a scenario where outputting to stderr here would be more useful.

It's outputting errors. Regardless of whether it's more useful: many scripts now assume the current behaviour so your suggestion would break these.
